### PR TITLE
Hit damage modifier

### PR DESCRIPTION
**Pull request by valdisz**
It is now possible to specify how many hits each successful attack will take.

The new parameter is set to `1` which guarantees not changes in behavior unless overridden in a specific ruleset.

To avoid marge conflicts, #92 must be merged first.

### CODE REVIEW

File: army.cpp [hunk 1]
```diff
@@ -1009,9 +1027,13 @@ Soldier * Army::GetAttacker(int i,int &behind)
 	return retval;
 }
 
-int Army::GetTargetNum(char const *special)
+int Army::GetTargetNum(char const *special, int canAttackBehind)
```
Comment (arkadybabaev) [inline army.cpp]: probably `bool` will be cleaner?

File: army.cpp [hunk 2]
```diff
@@ -1271,16 +1333,22 @@ int Army::DoAnAttack(Battle * b, char const *special, int numAttacks, int attack
...
 	Soldier *temp = soldiers[killed];
 
 	if (temp->amuletofi) return;
 
 	if (Globals->ARMY_ROUT == GameDefs::ARMY_ROUT_HITS_INDIVIDUAL)
 		hitsalive--;
-	temp->hits--;
-	temp->damage++;
+
+	int hitsLeft = temp->hits - damage;
```
Comment (arkadybabaev) [inline army.cpp]: Probably instead of lines 1345-1351 it will be cleaner to use:

`    temp->damage += std::min(temp->hits, damage);`
`    temp->hits = std::max(0, temp->hits - damage);`

File: army.h [hunk 3]
```diff
@@ -119,18 +120,20 @@ class Army
...
 		int NumAlive();
+		int NumBehind();
 		int NumSpoilers();
 		int CanAttack();
 		int NumFront();
+		int NumFrontHits();
 		Soldier *GetAttacker( int, int & );
 		int GetEffectNum(char const *effect);
-		int GetTargetNum(char const *special = NULL);
+		int GetTargetNum(char const *special = NULL, int canAttackBehind = 0);
```
Comment (arkadybabaev) [inline army.h]: probably cleaner is `bool` initialized with false.

File: battle.cpp [hunk 4]
```diff
@@ -42,6 +42,24 @@ Battle::~Battle()
...
+	if (Globals->ARMY_ROUT == GameDefs::ARMY_ROUT_FIGURES) {
+		aPower = Globals->OVERWHELMING * a->NumFront();
+		bPower = b->NumFront();
+	}
+	else {
+		aPower = Globals->OVERWHELMING * a->NumFrontHits();
+		bPower = b->NumFrontHits();
+	}
+
+	return bPower > aPower && a->NumFront() && a->NumBehind();
```
Comment (arkadybabaev) [inline battle.cpp]: It seems that `a->NumFront() && a->NumBehind()` may be redundant.

File: battle.cpp [hunk 5]
```diff
@@ -59,13 +77,18 @@ void Battle::FreeRound(Army * att,Army * def, int ass)
...
+	}
+
 	/* Run attacks until done */
 	int alv = def->NumAlive();
 	while (att->CanAttack() && def->NumAlive()) {
 		int num = getrandom(att->CanAttack());
 		int behind;
 		Soldier * a = att->GetAttacker(num, behind);
-		DoAttack(att->round, a, att, def, behind, ass);
+		DoAttack(att->round, a, att, def, behind, ass, overwhelmed);
```
Comment (arkadybabaev) [inline battle.cpp]: here we provide bool to int.

File: fracas/rules.cpp [hunk 6]
```diff
@@ -234,6 +234,7 @@ static GameDefs g = {
 	0,	// UNDEATH_CONTAGION
 	0,   // REGIONS_ECONOMY
 	0, // ADVANCED_TACTICS
+	0,	// OVERWHELMING
```
Comment (arkadybabaev) [inline fracas/rules.cpp]: probably its better to extend comment to explain feature. Otherwise it may be left as //REGIONS_ECONOMY and none actually knows what is it.

File: items.h [hunk 7]
```diff
@@ -207,6 +209,14 @@ enum {
 };
 
 
+// Describes bonus/mauls against another weapon
+class WeaponBonusMalus {
```
Comment (arkadybabaev) [inline items.h:217]: as all other my comments -- its just a comment.
But in all x64 size of the word is 8, and due to pointer in structure, it will be anyway aligned to 8.
So its nice to use `int32_t attackModifer; int32_t defenseModifer` so structure size will be 16, not 24.

File: army.cpp [hunk 8]
```diff
@@ -1134,9 +1162,24 @@ int Army::RemoveEffects(int num, char const *effect)
 	return(ret);
 }
 
+WeaponBonusMalus* GetWeaponBonusMalus(WeaponType *weapon, WeaponType *target) {
```
Comment (artyomtrityak) [inline army.cpp:1165]: i am 100% in favor of multi hit weapons but weapon bonus towards another weapon imho complicates things. Can you please extract it into separated PR as i would love to have multi damage weapons but still discuss this.

Comment (valdisz) [reply army.cpp:1165]: I will not have time to extract this into separate PR. Just do not add any bonuses. Also, there are missing methods to add bonuses for the ruleset, I'll add them later.

**Comment by valdisz** [comment]
This was the final commit and I consider the multi-hit damage feature completed. This code IS tested and it works as expected including weapon bonus-malus system.

**Comment by artyomtrityak** [comment]
Please change PR towards v3 branch once you have time!

**Comment by valdisz** [comment]
@artyomtrityak this is rebased

File: fracas/rules.cpp (hunk 6 above)
Comment (valdisz) [reply fracas/rules.cpp]: Description added in main file

File: items.h (hunk 7 above)
Comment (valdisz) [reply items.h:217]: I will not do that, mainly because I cannot predict what will outcome (not C++ expert) and 8 bytes per one item won't be noticeable as the weapon list is fixed and it is a singleton.

**Comment by valdisz** [comment]
@artyomtrityak I have rebased this branch. this resolved all merge conflicts

File: army.cpp [hunk 9]
```diff
@@ -26,6 +26,26 @@
...
+
+int max(const int a, const int b) {
+	return a > b ? a : b;
+}
+
+bool stringCompare(const char *str1, const char *str2) {
+    while (*str1 == *str2) {
+        if (*str1 == '\0' && *str2 == '\0') return true; 
+        
+		str1++;
```
Comment (artyomtrityak) [inline army.cpp]: why indentation here?

File: army.cpp [hunk 10]
```diff
@@ -1160,9 +1182,22 @@ int Army::RemoveEffects(int num, char const *effect)
 	return(ret);
 }
 
+WeaponBonusMalus* GetWeaponBonusMalus(WeaponType *weapon, WeaponType *target) {
```
Comment (artyomtrityak) [inline army.cpp:1165]: please add it to .h header file

File: army.cpp [hunk 11]
```diff
@@ -26,6 +26,26 @@
 #include "gameio.h"
 #include "gamedata.h"
 
+// as we are not using STD we need this...
+int min(const int a, const int b) {
```
Comment (artyomtrityak) [inline army.cpp]: please add them to .h file

**Comment by artyomtrityak** [comment]
Please try to use astring.cpp AString - it has all methods we need to compare.

File: army.cpp (hunk 11 above)
Comment (valdisz) [reply army.cpp]: added

File: battle.cpp [hunk 12]
```diff
@@ -57,7 +57,7 @@ bool IsArmyOverwhelmedBy(Army * a, Army * b) {
 		bPower = b->NumFrontHits();
 	}
 
-	return bPower > aPower && a->NumFront() && a->NumBehind();
+	return bPower > aPower;
```
Comment (artyomtrityak) [inline battle.cpp:60]: why this change? is there any issue with Overwhelming it fixes?

File: items.h [hunk 13]
```diff
@@ -209,6 +211,16 @@ enum {
 };
 
 
+#define MAX_WEAPON_BM_TARGETS	4
+
+// Describes bonus/mauls against another weapon
+class WeaponBonusMalus {
+	public:
+		char const *weaponAbbr;	// weapon abbreviation
+		int32_t attackModifer;		// how much increase/decrase attack versus this weapon
```
Comment (artyomtrityak) [inline items.h]: any reason we can not use just `int` here?

File: battle.cpp (hunk 12 above)
Comment (valdisz) [reply battle.cpp:60]: redundant

**Review by artyomtrityak — APPROVED** [review]
